### PR TITLE
Direct IO

### PR DESCRIPTION
**Pull request by zvin**
Major change

Only use aligned buffers to write to block devices.
It means all read streams read into aligned buffers when the target is a block device.
When we don't control the read stream, use a transform that will copy its buffers to aligned buffers.
Each read stream uses a pool of 1 MiB buffers to minimize allocations.

This should improve speed and reduce cpu usage.

### CODE REVIEW

File: lib/scanner/adapters/block-device.ts [hunk 1]
```diff
@@ -59,7 +59,12 @@ export class BlockDeviceAdapter extends Adapter {
 	private running = false;
 	private ready = false;
 
-	constructor(public includeSystemDrives: () => boolean = () => false) {
+	constructor(
+		public includeSystemDrives: () => boolean = () => false,
+		private unmountOnSuccess = false,
+		private oWrite = false,
+		private oDirect = true,
```
Comment (Page-) [inline lib/scanner/adapters/block-device.ts]: Can you make these an options object please, because something like `new scanner.adapters.BlockDeviceAdapter(() => false, false, true, true),` is very hard to understand

File: lib/aligned-lockable-buffer.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	buffer.alignment = alignment;
+	buffer.lock = mutex.lock.bind(mutex);
+	buffer.rlock = mutex.rlock.bind(mutex);
+	const bufferSlice = buffer.slice.bind(buffer);
+	buffer.slice = (...args) => {
+		const slice = bufferSlice(...args);
+		return attachMutex(slice, mutex, alignment);
+	};
+	return buffer;
+}
```
Comment (Page-) [inline lib/aligned-lockable-buffer.ts]: This could benefit from the same improvements I mentioned for a standard lockable buffer

File: lib/aligned-lockable-buffer.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		getAlignedBuffer(size, alignment),
+		new RWMutex(),
+		alignment,
+	);
+}
+
+export function isAlignedLockableBuffer(
+	buffer: Buffer,
+): buffer is AlignedLockableBuffer {
+	return (buffer as AlignedLockableBuffer).rlock !== undefined;
```
Comment (Page-) [inline lib/aligned-lockable-buffer.ts]: You can do this type check without the cast using
```suggestion
	return 'rlock' in buffer;
```
or
```suggestion
	return 'rlock' in buffer && buffer.rlock !== undefined;
```

File: lib/block-transform-stream.ts [hunk 4]
```diff
@@ -16,60 +16,92 @@
...
+				this.inputBytes += block.length - length;
 				block = block.slice(0, length);
 			}
 
-			this.bytesWritten += block.length;
-			this.push(block);
+			const alignedBuffer = this.alignedReadableState.getCurrentBuffer();
+			const unlock = await alignedBuffer.lock();
+			block.copy(alignedBuffer);
+			unlock();
```
Comment (Page-) [inline lib/block-transform-stream.ts]: Should this be
```suggestion
			const unlock = await alignedBuffer.lock();
			try {
				block.copy(alignedBuffer);
			} finally {
				unlock();
			}
```
or can the copy never fail or not unlocking on failure is ok? (same for other cases like this)

File: lib/block-transform-stream.ts [hunk 5]
```diff
@@ -16,60 +16,92 @@
...
-			this.push(chunk);
-		} else {
-			this._buffers.push(chunk);
-			this._bytes += chunk.length;
-			this.writeBuffers();
-		}
+		this.inputBuffers.push(chunk);
+		this.inputBytes += chunk.length;
+		await this.writeBuffers();
 		callback();
```
Comment (Page-) [inline lib/block-transform-stream.ts]: Should there be a try/catch here in case `this.writeBuffers()` is rejected? (and the same for other places where promises/callbacks are mixed like this)

File: lib/block-transform-stream.ts [hunk 6]
```diff
@@ -16,60 +16,92 @@
...
+	) {
+		if (alignment === undefined) {
+			return stream;
+		}
+		const transform = new BlockTransformStream({
+			chunkSize: CHUNK_SIZE,
+			alignment,
+			numBuffers,
+		});
+		stream.pipe(transform);
```
Comment (Page-) [inline lib/block-transform-stream.ts]: Does this stream need an error handler?

File: lib/source-destination/multi-destination.ts [hunk 7]
```diff
@@ -108,9 +114,21 @@ export class MultiDestination extends SourceDestination {
...
+	public getAlignment(): number | undefined {
+		const alignments: number[] = [];
+		for (const destination of this.destinations.values()) {
+			if (destination instanceof BlockDevice) {
+				alignments.push(destination.alignment);
+			}
+		}
+		if (alignments.length) {
+			return Math.max(...alignments);
+		}
```
Comment (Page-) [inline lib/source-destination/multi-destination.ts]: This can be
```suggestion
		return _.maxBy(this.destinations.values(), () => {
			if (destination instanceof BlockDevice) {
				return destination.alignment;
			}
		});
```

File: lib/lockable-buffer.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+	const bufferSlice = buffer.slice.bind(buffer);
+	buffer.slice = (...args) => {
+		const slice = bufferSlice(...args);
+		return attachMutex(slice, mutex);
+	};
+	return buffer;
+}
+
+export function createBuffer(size: number, alignment: number): LockableBuffer {
+	return attachMutex(getAlignedBuffer(size, alignment), new RWMutex());
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/lockable-buffer.ts]: I ended up with this @Page- 
```typescript
function alignedLockableBufferSlice(
	this: AlignedLockableBuffer,
	start?: number,
	end?: number,
) {
	const slice = Buffer.prototype.slice.call(this, start, end);
	return attachMutex(slice, this.alignment, this.lock, this.rlock);
}

function attachMutex(
	buf: Buffer,
	alignment: number,
	lock: () => Promise<() => void>,
	rlock: () => Promise<() => void>,
): AlignedLockableBuffer {
	const buffer = buf as AlignedLockableBuffer;
	buffer.alignment = alignment;
	buffer.lock = lock;
	buffer.rlock = rlock;
	buffer.slice = alignedLockableBufferSlice;
	return buffer;
}

export function createBuffer(
	size: number,
	alignment: number,
): AlignedLockableBuffer {
	const mutex = new RWMutex();
	return attachMutex(
		getAlignedBuffer(size, alignment),
		alignment,
		mutex.lock.bind(mutex),
		mutex.rlock.bind(mutex),
	);
}
```

File: lib/source-destination/multi-destination.ts (hunk 7 above)
Comment (zvin) [reply lib/source-destination/multi-destination.ts]: I didn't change this one as:
 * this would require adding an `Array.from` as lodash does not work with iterators;
 * `maxBy` would return the BlockDevice itself, not the alignment;

Comment (zvin) [reply lib/source-destination/multi-destination.ts]: replaced with 
```
	public getAlignment(): number | undefined {
		let max = undefined;
		for (const destination of this.destinations.values()) {
			if ((destination instanceof BlockDevice) && (max === undefined || destination.alignment > max)) {
				max = destination.alignment;
			}
		}
		return max;
	}
```

Comment (zvin) [reply lib/source-destination/multi-destination.ts]: Actually to be 100% correct, this should return the least common multiple of all alignments.
But as the only alignment I've seen are 512 and 4096, max also works.

File: lib/source-destination/file.ts [hunk 9]
```diff
@@ -135,6 +133,7 @@ export class File extends SourceDestination {
 	}
 
 	public async createWriteStream(): Promise<NodeJS.WritableStream> {
+		// @ts-ignore: @types/node is wrong about fs.WriteStream constructor: it takes 2 arguments, the first one is the file path
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/source-destination/file.ts]: Why ?
`createWriteStream()` returns a `ProgressWriteStream` which is an `fs.WriteStream` with some stuff added.

File: lib/source-destination/multi-destination.ts (hunk 7 above)
Comment (Page-) [reply lib/source-destination/multi-destination.ts]: You could possibly start with `let max = 0;` and avoid the `undefined` check

File: lib/lockable-buffer.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Page-) [reply lib/lockable-buffer.ts]: Looks good

File: lib/source-destination/multi-destination.ts (hunk 7 above)
Comment (zvin) [reply lib/source-destination/multi-destination.ts]: A 0 `alignment` would make no sense.

File: lib/source-destination/file.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Page-) [reply lib/source-destination/file.ts]: Well the ProgressWriteStream surely has the right types doesn't it @zvin? In which case the `@ts-ignore` here is unnecessary and could potentially hide issues

File: lib/source-destination/multi-destination.ts (hunk 7 above)
Comment (zvin) [reply lib/source-destination/multi-destination.ts]: I prefer keeping it as `undefined` => don't align.

Comment (Page-) [reply lib/source-destination/multi-destination.ts]: You could do it as `if (alignment === 0) return undefined` but I'm fine with the version you posted in that case as well

File: lib/source-destination/file.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (zvin) [reply lib/source-destination/file.ts]: It does not as it is not redefining the constructor.

[the remark this answers is not in the document]
Comment (Page-) [reply lib/source-destination/file.ts]: Ok, I did just see in ProgressWriteStream:
```
export const ProgressWriteStream = makeClassEmitProgressEvents(
	// type definitions for node 6 export fs.WriteStream as an interface, but it's a class.
	// @ts-ignore
	fs.WriteStream,
	'bytesWritten',
	'bytesWritten',
	PROGRESS_EMISSION_INTERVAL,
);
```
it mentions an issue with node 6 typings - that can probably be updated now since you updated the node typings, it might also be worth checking if the node 8 typings fix this issue

[the remark this answers is not in the document]
Comment (zvin) [reply lib/source-destination/file.ts]: They're still wrong in node 13 types: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/fs.d.ts#L164

The comment was updated in another commit, it no longer mentions node 6.

[the remark this answers is not in the document]
Comment (zvin) [reply lib/source-destination/file.ts]: The comment about interface vs class has been removed as this one is fixed.

[the remark this answers is not in the document]
Comment (zvin) [reply lib/source-destination/file.ts]: Actual implementation here https://github.com/nodejs/node/blob/master/lib/internal/fs/streams.js#L286

File: lib/aligned-lockable-buffer.ts (hunk 3 above)
Comment (zvin) [reply lib/aligned-lockable-buffer.ts]: updated

File: lib/block-transform-stream.ts (hunk 4 above)
Comment (zvin) [reply lib/block-transform-stream.ts]: updated

File: lib/block-transform-stream.ts (hunk 5 above)
Comment (zvin) [reply lib/block-transform-stream.ts]: added

File: lib/block-transform-stream.ts (hunk 6 above)
Comment (zvin) [reply lib/block-transform-stream.ts]: added

File: lib/source-destination/dmg.ts [hunk 10]
```diff
@@ -69,15 +78,29 @@ export class DmgSource extends SourceSource {
...
+			blocks,
 		});
+		if (alignment !== undefined) {
+			const transform = new ProgressSparseTransformStream({
+				blocks,
+				chunkSize: CHUNK_SIZE,
+				alignment,
+				numBuffers,
+			});
+			stream.pipe(transform);
```
Comment (Page-) [inline lib/source-destination/dmg.ts]: Should `stream` have an error handler?

File: lib/block-transform-stream.ts [hunk 11]
```diff
@@ -16,60 +16,95 @@
...
-		callback();
+		this.inputBuffers.push(chunk);
+		this.inputBytes += chunk.length;
+		asCallback(this.writeBuffers(), callback);
 	}
 
-	public _flush(callback: (error?: Error) => void) {
-		this.writeBuffers(true);
-		callback();
+	public async _flush(callback: (error?: Error) => void): Promise<void> {
```
Comment (Page-) [inline lib/block-transform-stream.ts]: This returns an immediately-resolved promise, is that intended or should this be
```suggestion
	public _flush(callback: (error?: Error) => void): void {
```
or have an await?

File: lib/block-transform-stream.ts [hunk 12]
```diff
@@ -16,60 +16,95 @@
...
 		}
 	}
 
-	public _transform(
+	public async _transform(
 		chunk: Buffer,
 		_encoding: string,
 		callback: (error?: Error) => void,
-	) {
+	): Promise<void> {
```
Comment (Page-) [inline lib/block-transform-stream.ts]: This is also an immediately-resolved promise

File: lib/block-read-stream.ts [hunk 13]
```diff
@@ -65,34 +89,30 @@ export class BlockReadStream extends Readable {
...
+		let buffer = this.alignedReadableState.getCurrentBuffer();
+		const toRead = this.end - this.bytesRead + 1;
+		if (toRead < buffer.length) {
+			buffer = buffer.slice(0, toRead);
+		}
 		try {
+			const unlock = await buffer.lock();
 			const { bytesRead } = await this.tryRead(buffer);
-			if (bytesRead === 0) {
+			unlock();
```
Comment (Page-) [inline lib/block-read-stream.ts]: Should this be in a `finally` block?

File: lib/sparse-stream/shared.ts [hunk 14]
```diff
@@ -55,7 +55,7 @@ export interface SparseWritable extends NodeJS.WritableStream {
 	_write(
 		chunk: SparseStreamChunk,
 		encoding: string,
-		callback: (err?: Error | void) => void,
+		callback: (err?: Error | void | null) => void,
```
Comment (Page-) [inline lib/sparse-stream/shared.ts]: I think this should be
```suggestion
		callback: (err?: Error | null) => void,
```
as void doesn't really make sense as a parameter value

File: lib/utils.ts [hunk 15]
```diff
@@ -84,7 +84,7 @@ export function difference<T>(setA: Set<T>, setB: Set<T>): Set<T> {
 
 export async function asCallback<T>(
 	promise: Promise<T>,
-	callback: (error: Error | void, value?: T) => void,
+	callback: (error: Error | void | null, value?: T) => void,
```
Comment (Page-) [inline lib/utils.ts]: Another place to drop `void`
```suggestion
	callback: (error: Error | null, value?: T) => void,
```

File: tests/zip.spec.ts [hunk 16]
```diff
@@ -78,7 +78,7 @@ describe('zip support', function() {
 	);
 
 	describe('compression method', function() {
-		const compressionMethods = [
+		const compressionMethods: string[] = [
```
Comment (Page-) [inline tests/zip.spec.ts]: Is this change actually necessary? I don't mind it but generally I leave typescript to do the inference as it makes changes a bit easier

File: tsconfig.json [hunk 17]
```diff
@@ -5,21 +5,14 @@
 		"outDir": "build",
 		"sourceMap": true,
 		"declaration": true,
-		"target": "es6",
+		"target": "es2017",
```
Comment (Page-) [inline tsconfig.json]: This could be updated to
```suggestion
		"target": "es2018",
```
with dropped node 8 support, the only thing node 10 doesn't support is `Proxy "ownKeys" handler, duplicate keys for non-extensible targets`: https://node.green/#ES2018

Comment (zvin) [reply tsconfig.json]: updated

File: tests/zip.spec.ts (hunk 16 above)
Comment (zvin) [reply tests/zip.spec.ts]: debugging leftover, removed

File: lib/sparse-stream/sparse-write-stream.ts [hunk 18]
```diff
@@ -62,31 +68,53 @@ export class SparseWriteStream extends Writable implements SparseWritable {
...
-		// This is to prevent Windows from mounting the device while we flash it.
-		if (chunk.position < this.firstBytesToKeep) {
-			const end = chunk.position + chunk.buffer.length;
-			if (end <= this.firstBytesToKeep) {
-				this._firstChunks.push(chunk);
-				this.position = chunk.position + chunk.buffer.length;
-				this.bytesWritten += chunk.buffer.length;
+		const unlock = isAlignedLockableBuffer(chunk.buffer)
+			? await chunk.buffer.rlock()
+			: noop;
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/sparse-stream/sparse-write-stream.ts]: updated

File: lib/sparse-stream/sparse-transform-stream.ts [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+				buffer = buffer.slice(0, chunk.buffer.length);
+			}
+			chunk.buffer.copy(buffer);
+		} finally {
+			unlock();
+		}
+		this.push({ position: chunk.position, buffer });
+		this.bytesWritten += chunk.buffer.length;
+		this.position += chunk.buffer.length;
+		callback();
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/sparse-stream/sparse-transform-stream.ts]: yes it should be handled, updated

File: lib/sparse-stream/sparse-read-stream.ts [hunk 20]
```diff
@@ -71,15 +98,23 @@ export class SparseReadStream extends Readable implements SparseReadable {
...
+		try {
+			await this.source.read(
+				buffer,
+				0,
+				length,
+				this.state.subBlock.offset + this.positionInBlock,
+			);
+			if (this.state.hasher !== undefined) {
+				this.state.hasher.update(buffer);
+			}
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/sparse-stream/sparse-read-stream.ts]: updated

File: lib/sparse-stream/sparse-read-stream.ts [hunk 21]
```diff
@@ -71,15 +98,23 @@ export class SparseReadStream extends Readable implements SparseReadable {
...
-			length,
-			this.state.subBlock.offset + this.positionInBlock,
-		);
-		if (this.state.hasher !== undefined) {
-			this.state.hasher.update(buffer);
+		const buffer =
+			this.alignedReadableState !== undefined
+				? this.alignedReadableState.getCurrentBuffer().slice(0, length)
+				: Buffer.allocUnsafe(length);
+		const unlock = isAlignedLockableBuffer(buffer) ? await buffer.lock() : noop;
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/sparse-stream/sparse-read-stream.ts]: updated

File: lib/source-destination/multi-destination.ts [hunk 22]
```diff
@@ -301,12 +323,16 @@ export class MultiDestination extends SourceDestination {
 		return passthrough;
 	}
 
-	public async createWriteStream(): Promise<NodeJS.WritableStream> {
-		return await this.createStream('createWriteStream');
+	public async createWriteStream(
+		...args: any[]
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/source-destination/multi-destination.ts]: nice, thanks

File: lib/source-destination/dmg.ts (hunk 10 above)
Comment (zvin) [reply lib/source-destination/dmg.ts]: added

File: lib/block-transform-stream.ts (hunk 11 above)
Comment (zvin) [reply lib/block-transform-stream.ts]: right, fixed

File: lib/block-transform-stream.ts (hunk 12 above)
Comment (zvin) [reply lib/block-transform-stream.ts]: fixed

File: lib/block-read-stream.ts (hunk 13 above)
Comment (zvin) [reply lib/block-read-stream.ts]: yes, added

File: lib/sparse-stream/shared.ts (hunk 14 above)
Comment (zvin) [reply lib/sparse-stream/shared.ts]: removed

File: lib/utils.ts (hunk 15 above)
Comment (zvin) [reply lib/utils.ts]: dropped

File: lib/utils.ts [hunk 23]
```diff
@@ -40,7 +52,16 @@ export async function sparseStreamToBuffer(
 	await new Promise((resolve: () => void, reject: (error: Error) => void) => {
 		stream.on('error', reject);
 		stream.on('end', resolve);
-		stream.on('data', chunks.push.bind(chunks));
+		stream.on('data', async (chunk: SparseStreamChunk) => {
+			if (isAlignedLockableBuffer(chunk.buffer)) {
+				const unlock = await chunk.buffer.rlock();
+				const data = Buffer.allocUnsafe(chunk.buffer.length);
+				chunk.buffer.copy(data);
+				unlock();
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/utils.ts]: good catch, added

File: lib/utils.ts [hunk 24]
```diff
@@ -40,7 +52,16 @@ export async function sparseStreamToBuffer(
 	await new Promise((resolve: () => void, reject: (error: Error) => void) => {
 		stream.on('error', reject);
 		stream.on('end', resolve);
-		stream.on('data', chunks.push.bind(chunks));
+		stream.on('data', async (chunk: SparseStreamChunk) => {
+			if (isAlignedLockableBuffer(chunk.buffer)) {
+				const unlock = await chunk.buffer.rlock();
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/utils.ts]: Right, this function is only used in tests, fixed it anyway.

File: lib/utils.ts [hunk 25]
```diff
@@ -23,7 +24,18 @@ export async function streamToBuffer(
...
 			const chunks: Buffer[] = [];
 			stream.on('error', reject);
-			stream.on('data', chunks.push.bind(chunks));
+			stream.on('data', async (chunk: Buffer) => {
+				let data: Buffer;
+				if (isAlignedLockableBuffer(chunk)) {
+					const unlock = await chunk.rlock();
+					data = Buffer.allocUnsafe(chunk.length);
+					chunk.copy(data);
+					unlock();
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/utils.ts]: added

File: lib/source-destination/multi-destination.ts [hunk 26]
```diff
@@ -217,23 +236,26 @@ export class MultiDestination extends SourceDestination {
...
 		return await Array.from(this.activeDestinations)[0].createSparseReadStream(
-			...args,
+			options,
 		);
 	}
 
 	private async createStream(
 		methodName: 'createWriteStream' | 'createSparseWriteStream',
+		...args: any[]
 	) {
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/source-destination/multi-destination.ts]: added, thanks

File: lib/block-transform-stream.ts [hunk 27]
```diff
@@ -16,34 +16,62 @@
...
+			numBuffers,
+		);
 	}
 
-	private writeBuffers(flush = false) {
-		if (flush || this._bytes >= this.chunkSize) {
-			let block = Buffer.concat(this._buffers);
+	private async writeBuffers(flush = false): Promise<void> {
+		if (flush || this.inputBytes >= this.chunkSize) {
+			// TODO optimize
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/block-transform-stream.ts]: I think we can probably completely avoid `Buffer.concat`, but that'll make it more complex to read.

File: lib/block-read-stream.ts [hunk 28]
```diff
@@ -65,34 +89,34 @@ export class BlockReadStream extends Readable {
...
+		if (toRead < buffer.length) {
+			buffer = buffer.slice(0, toRead);
+		}
 		try {
-			const { bytesRead } = await this.tryRead(buffer);
-			if (bytesRead === 0) {
-				this.push(null);
-				return;
+			const unlock = await buffer.lock();
+			let bytesRead: number;
```
[the remark this answers is not in the document]
Comment (zvin) [reply lib/block-read-stream.ts]: updated
